# DI 4.9 - Part I Task

## Importing Libraries

In [5]:
# Import Libraries

import pandas as pd
import numpy as np
import os

In [9]:
# Import New Customer Data Frame

df_cust = pd.read_csv(r'D:\Career Foundry\2) Data Immersion\4) Python Fundamentals for Data Analysts\DI 4.2 - Jupyter Folder\Instacart Basket Analysis\02 Data\Original Data\customers data 4.9\customers.csv', index_col = False)

## Exploring Customer Data

In [49]:
df_cust.describe()

user_id            Age   independants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [14]:
df_cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [18]:
df_cust.shape

(206209, 10)

In [24]:
df_cust.head(20)

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0     26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1     33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2     65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3    125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4    130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5    133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6    152052      Chris    Walton    Male         Montana   20    1/1/2017   
7    168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8     69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9     82820      Shawn     Chung    Male        Virginia   26    1/1/2017   
10    26441     Gloria  Stafford  Female          Nevada   19    1/1/2017   
11    33835      Roger     Casey    Male        Nebraska   45    1/1/2017   
12    75032    Cynthia  Mcdaniel  Female         Georgia   75    1/1/2017   
13       51  Stephanie    Jensen  Female         Wyoming   37    1/1/2017   
14      516      Peter      Hunt    Male        Colorado   51    1/1/2017   
15     6965    Brandon      Mata    Male    North Dakota   70    1/1/2017   
16     8794        Amy      Mack  Female        Nebraska   30    1/1/2017   
17     9944      Brian    Oneill    Male       Wisconsin   26    1/1/2017   
18    13313     Victor      Moss    Male          Alaska   31    1/1/2017   
19    17335    Patrick    Fowler    Male         Vermont   65    1/1/2017   

    n_dependants                        fam_status  income  
0              3                           married  165665  
1              0                            single   59285  
2              2                           married   99568  
3              0                            single   42049  
4              1                           married   40374  
5              2                           married   49643  
6              0                            single   61746  
7              0                            single   63712  
8              1                           married  162432  
9              2                           married   32072  
10             2  living with parents and siblings   43443  
11             2                           married   47988  
12             1                           married   57835  
13             1                           married   71190  
14             2                           married  146559  
15             0                  divorced/widowed  100091  
16             3                           married   57096  
17             1                           married   91826  
18             2                           married   91111  
19             1                           married   34514

In [26]:
df_cust.tail(20)

user_id   First Name      Surnam  Gender           STATE  Age  \
206189   193828      Russell      Travis    Male  North Carolina   46   
206190   197067        Kathy        Bell  Female         Arizona   42   
206191     7177      Russell   Zimmerman    Male     Mississippi   71   
206192    61888       Joshua      Guerra    Male      New Jersey   37   
206193   103412       Willie     Goodman    Male        Michigan   46   
206194   189337        Shawn        Wood    Male      New Jersey   56   
206195   205766       Amanda       Hodge  Female          Oregon   18   
206196   139950       Gloria      Murray  Female        Colorado   45   
206197    74598  Christopher   Velazquez    Male       Minnesota   52   
206198    83573       Gloria      Murray  Female        Michigan   28   
206199   179673         Adam  Villanueva    Male         Wyoming   77   
206200   192637        Kelly    Petersen  Female         Florida   47   
206201    75529        Kathy       Avila  Female   West Virginia   25   
206202   157533          Ned       Stark    Male            Utah   53   
206203   122741         Carl       Weber    Male        Michigan   24   
206204   168073         Lisa        Case  Female  North Carolina   44   
206205    49635       Jeremy     Robbins    Male          Hawaii   62   
206206   135902        Doris    Richmond  Female        Missouri   66   
206207    81095         Rose     Rollins  Female      California   27   
206208    80148      Cynthia       Noble  Female        New York   55   

       date_joined  n_dependants                        fam_status  income  
206189    4/1/2020             1                           married  160483  
206190    4/1/2020             0                            single  114821  
206191    4/1/2020             3                           married   64400  
206192    4/1/2020             1                           married   68491  
206193    4/1/2020             2                           married  154481  
206194    4/1/2020             1                           married   40373  
206195    4/1/2020             2  living with parents and siblings   48510  
206196    4/1/2020             2                           married  150954  
206197    4/1/2020             0                            single  140700  
206198    4/1/2020             0                            single   32237  
206199    4/1/2020             0                  divorced/widowed  162239  
206200    4/1/2020             0                            single   50638  
206201    4/1/2020             2                           married   56513  
206202    4/1/2020             0                            single   35973  
206203    4/1/2020             2                           married   73357  
206204    4/1/2020             1                           married  148828  
206205    4/1/2020             3                           married  168639  
206206    4/1/2020             2                           married   53374  
206207    4/1/2020             1                           married   99799  
206208    4/1/2020             1                           married   57095

## Wrangling (renaming and changing data types)

In [47]:
# Changing column names (Surnam to Last Name)

df_cust.rename(columns = {'Surnam' : 'Last Name'}, inplace = True)

In [33]:
# Changing column names (STATE to State)

df_cust.rename(columns = {'STATE' : 'State'}, inplace = True)

In [45]:
# Changing column names (n_dependants to dependants)

df_cust.rename(columns = {'n_dependants' : 'dependants'}, inplace = True)

In [41]:
# Checking Changes were Successful

df_cust.head()

user_id First Name Last Name  Gender       State  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   independants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [122]:
# Changing user_id to Integer

df_cust['user_id'] = df_cust['user_id'].astype('int')

In [124]:
# Checking Changes were Successful (pt1)

df_cust['user_id'].dtype

dtype('int32')

In [126]:
# Checking Changes were Successful (pt2)

df_cust.describe()

user_id            Age   independants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [128]:
# Checking Changes were Successful (pt3)

df_cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int32 
 1   First Name    206209 non-null  object
 2   Last Name     206209 non-null  object
 3   Gender        206209 non-null  object
 4   State         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   independants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int32(1), int64(3), object(6)
memory usage: 14.9+ MB


##### No column dropping as in Part II of this task it appears all columns will be needed for analysis. No data transposing needed.

## Data Consistency Checks

In [73]:
# Checking for Mixed type data

for col in df_cust.columns.tolist():
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust[weird]) > 0:
    print (col)

First Name


C:\Users\jmesi\AppData\Local\Temp\ipykernel_158096\1686986452.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\jmesi\AppData\Local\Temp\ipykernel_158096\1686986452.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\jmesi\AppData\Local\Temp\ipykernel_158096\1686986452.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\jmesi\AppData\Local\Temp\ipykernel_158096\1686986452.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\jmesi\AppData\L

In [75]:
# Correcting First Name data type to str

df_cust['First Name'] = df_cust['First Name'].astype('str')

In [79]:
# Checking changes were successful (pt1)

df_cust['First Name'].dtype

dtype('O')

In [85]:
# Checking changes were successful (pt2)

df_cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  object
 1   First Name    206209 non-null  object
 2   Last Name     206209 non-null  object
 3   Gender        206209 non-null  object
 4   State         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   independants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(3), object(7)
memory usage: 15.7+ MB


##### First Name column Non-Null numerical value went from 194950 (ref. to Wrangled section) to 206209. That means 11,259 First Name entries were Mixed Type Data

In [87]:
# Checking for missing values

df_cust.isnull().sum()

user_id         0
First Name      0
Last Name       0
Gender          0
State           0
Age             0
date_joined     0
independants    0
fam_status      0
income          0
dtype: int64

##### No missing values detected

In [96]:
# Checking for duplicate data

df_cust_dups = df_cust[df_cust.duplicated()]

In [98]:
df_cust_dups

Empty DataFrame
Columns: [user_id, First Name, Last Name, Gender, State, Age, date_joined, independants, fam_status, income]
Index: []

##### No duplicate data detected

## Combining Customer Data with Prepared Instacart Data

In [104]:
# Importing Prepared Instacart Data

ords_prods_merge = pd.read_pickle(r'D:\Career Foundry\2) Data Immersion\4) Python Fundamentals for Data Analysts\DI 4.2 - Jupyter Folder\Instacart Basket Analysis\02 Data\Prepared Data\ords_prods_merge_4.9.pkl')

In [137]:
# Exploring ords_prods_merge Data

ords_prods_merge

Unnamed: 0_x  product_id                product_name  aisle_id  \
0                    0           1  Chocolate Sandwich Cookies        61   
1                    0           1  Chocolate Sandwich Cookies        61   
2                    0           1  Chocolate Sandwich Cookies        61   
3                    0           1  Chocolate Sandwich Cookies        61   
4                    0           1  Chocolate Sandwich Cookies        61   
...                ...         ...                         ...       ...   
32404854         49692       49688      Fresh Foaming Cleanser        73   
32404855         49692       49688      Fresh Foaming Cleanser        73   
32404856         49692       49688      Fresh Foaming Cleanser        73   
32404857         49692       49688      Fresh Foaming Cleanser        73   
32404858         49692       49688      Fresh Foaming Cleanser        73   

          department_id  prices  Unnamed: 0_y  order_id  user_id eval_set  \
0                    19     5.8          1987   3139998      138    prior   
1                    19     5.8          1989   1977647      138    prior   
2                    19     5.8         11433    389851      709    prior   
3                    19     5.8         12198    652770      764    prior   
4                    19     5.8         12200   1813452      764    prior   
...                 ...     ...           ...       ...      ...      ...   
32404854             11    13.5       3321928   1788356   200215    prior   
32404855             11    13.5       3324686   3401313   200377    prior   
32404856             11    13.5       3332310    809510   200873    prior   
32404857             11    13.5       3332314   2359893   200873    prior   
32404858             11    13.5       3416620   2385091   205926    prior   

          ...  _merge  merge_source    price_range_loc     Busiest day  \
0         ...    both          both  Mid-range product  Regularly busy   
1         ...    both          both  Mid-range product  Regularly busy   
2         ...    both          both  Mid-range product    Busiest days   
3         ...    both          both  Mid-range product  Regularly busy   
4         ...    both          both  Mid-range product      Least busy   
...       ...     ...           ...                ...             ...   
32404854  ...    both          both                NaN    Busiest days   
32404855  ...    both          both                NaN      Least busy   
32404856  ...    both          both                NaN  Regularly busy   
32404857  ...    both          both                NaN  Regularly busy   
32404858  ...    both          both                NaN  Regularly busy   

            Busiest days  busiest_period_of_day max_order      loyalty_flag  \
0         Regularly busy            Most orders        32  Regular customer   
1         Regularly busy         Average orders        32  Regular customer   
2           Busiest days         Average orders         5      New customer   
3             Least busy            Most orders         3      New customer   
4             Least busy         Average orders         3      New customer   
...                  ...                    ...       ...               ...   
32404854    Busiest days            Most orders         6      New customer   
32404855      Least busy            Most orders         4      New customer   
32404856      Least busy         Average orders        20  Regular customer   
32404857      Least busy            Most orders        20  Regular customer   
32404858    Busiest days            Most orders        11  Regular customer   

          spender_type               freq  
0          Low spender  Frequent customer  
1          Low spender   Regular customer  
2          Low spender  Frequent customer  
3          Low spender          10.948718  
4          Low spender  Frequent customer  
...                ...                ...  
32404854  High spender  Fr

In [106]:
# Exploring ords_prods_merge Data

ords_prods_merge.shape

(32404859, 26)

In [110]:
# Exploring ords_prods_merge Data

ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 26 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   Unnamed: 0_x            int64   
 1   product_id              int64   
 2   product_name            object  
 3   aisle_id                int64   
 4   department_id           int64   
 5   prices                  float64 
 6   Unnamed: 0_y            int64   
 7   order_id                int64   
 8   user_id                 int64   
 9   eval_set                object  
 10  order_number            int64   
 11  order_dow               int64   
 12  order_hour_of_day       int64   
 13  days_since_prior_order  float64 
 14  add_to_cart_order       int64   
 15  reordered               int64   
 16  _merge                  category
 17  merge_source            category
 18  price_range_loc         object  
 19  Busiest day             object  
 20  Busiest days            object  
 21  busies

In [132]:
# Combinging ords_prods_merge with df_cust

df_merged_large = pd.merge(ords_prods_merge, df_cust, on = 'user_id', how='inner', indicator= 'merge_origin') 

##### I had to go back into my Wrangle section and change the 'user_id' from str to int. As I was trying to merge the two df I was getting errors that the dtype were not matching. This has reintroduced the 'user_id' back into the describe function, giving me statistical data that's not necessary.

In [145]:
# Checking the Inner Merge was Successful

df_merged_large.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 36 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   Unnamed: 0_x            int64   
 1   product_id              int64   
 2   product_name            object  
 3   aisle_id                int64   
 4   department_id           int64   
 5   prices                  float64 
 6   Unnamed: 0_y            int64   
 7   order_id                int64   
 8   user_id                 int64   
 9   eval_set                object  
 10  order_number            int64   
 11  order_dow               int64   
 12  order_hour_of_day       int64   
 13  days_since_prior_order  float64 
 14  add_to_cart_order       int64   
 15  reordered               int64   
 16  _merge                  category
 17  merge_source            category
 18  price_range_loc         object  
 19  Busiest day             object  
 20  Busiest days            object  
 21  busies

In [147]:
df_merged_large.shape

(32404859, 36)

##### In the ords_prods_merge file the shape was: (32404859, 26). Once merged with the df_cust that had a shape of: (206209, 10) the total shape became: (32404859, 36). The amount of rows did not change but the amount of columns did.

## Exporting file as Pickle

In [152]:
path=r'D:\Career Foundry\2) Data Immersion\4) Python Fundamentals for Data Analysts\DI 4.2 - Jupyter Folder\Instacart Basket Analysis'

In [ ]:
df_merged_large.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'df_merged_large_pt4.9.pkl'))